### PR TITLE
oneDNN support for input memory format.

### PR DESCRIPTION
**Pull request by arogowie-intel**
### PR types
New features

### PR changes
OPs, APIs

### Describe
This PR introduces `FLAGS_onednn_use_input_mem_format` gflag flag as well as appropriate API for `AnalysisConfig`:
Python:
```py
config = fluid.core.AnalysisConfig(...)
config.enable_onednn_use_input_mem_format()
```
C++
```c++
AnalysisConfig cfg;
cfg.EnableOneDNNUseInputMemFormat();
```

This flag decides whether to use input tensor memory format as is. By default for compute-intensive operators oneDNN kernels use it's own optimized memory format. However this usually results in additional reorders and copying tensor data, thus more memory consumption. With this flag set to true oneDNN kernels will use input tensor's Paddle memory format, which will result in less memory consumption, although at a cost of performance. 

At this moment this support is added to Conv2d and FC operators.

For Resnet50 we can see around 10-14% less memory consumption with around 2x slow-down.

This feature is the beginning for future development of memory optimizations when oneDNN library is used.

### CODE REVIEW

**Comment by arogowie-intel** [comment]
@jczaja @wojtuss @lidanqing-intel @arlesniak Could you please review?

**Review by jczaja — DISMISSED** [review]
LGTM

**Comment by arogowie-intel** [comment]
Hi @luotao1 Could you advise on how to link `analyzer_image_classification_tester.cc` with `flags.cc` file where I have defined `onednn_use_input_mem_format` flag? I've tried to add `flags` target to the list of extra deps in CMakefile. Although this worked for me locally and in majority of CI, some CI checks fail with following errors.

```
CMakeFiles/test_analyzer_image_classification.dir/analyzer_image_classification_tester.cc.o: In function `paddle::inference::analysis::compare(bool)':
/paddle/paddle/fluid/inference/tests/api/analyzer_image_classification_tester.cc:90: undefined reference to `fLB::FLAGS_onednn_use_input_mem_format'
CMakeFiles/test_analyzer_image_classification.dir/analyzer_image_classification_tester.cc.o: In function `paddle::inference::analysis::profile(bool)':
/paddle/paddle/fluid/inference/tests/api/analyzer_image_classification_tester.cc:54: undefined reference to `fLB::FLAGS_onednn_use_input_mem_format'
```

File: paddle/fluid/inference/tests/api/CMakeLists.txt [hunk 1]
```diff
@@ -105,9 +105,10 @@ function(inference_analysis_api_object_dection_int8_test_run TARGET_NAME test_bi
 	     --iterations=1)
 endfunction()
 
+# (aosewski) link to 'flags' to get access to 'onednn_use_input_mem_format' global flag.
 function(inference_analysis_api_test_with_fake_data_build TARGET_NAME filename)
 	inference_analysis_test_build(${TARGET_NAME} SRCS ${filename}
-        EXTRA_DEPS ${INFERENCE_EXTRA_DEPS})
+        EXTRA_DEPS ${INFERENCE_EXTRA_DEPS} flags)
 endfunction()
```
Comment (luotao1) [inline paddle/fluid/inference/tests/api/CMakeLists.txt]: > Could you advise on how to link analyzer_image_classification_tester.cc with flags.cc file where I have defined onednn_use_input_mem_format flag? I've tried to add flags target to the list of extra deps in CMakefile. Although this worked for me locally and in majority of CI, some CI checks fail with following errors.

I think you should not add flags target to the the list of extra deps in CMakefile. If you want to pass CI, how about ? 
```
void profile(bool use_mkldnn = false, bool onednn_use_input_mem_format = false) {
  if (use_mkldnn) {
      cfg.EnableMKLDNN();
      if (onednn_use_input_mem_format) 
        cfg.EnableOneDNNUseInputMemFormat();
    ....
  }
}

#ifdef PADDLE_WITH_MKLDNN
TEST(Analyzer_resnet50, profile_mkldnn) { profile(true /* use_mkldnn */, false /* onednn_use_input_mem_format */); }
TEST(Analyzer_resnet50, profile_mkldnn) { profile(true /* use_mkldnn */, true /* onednn_use_input_mem_format */); }
#endif
```

**Comment by arogowie-intel** [comment]
@luotao1 Could you advise why PR-CI-Coverage complains about:
```
File "/paddle/build/python/paddle/fluid/tests/unittests/mkldnn/test_onednn_use_input_mem_format.py", line 23, in <module>
     from paddle.fluid.tests.unittests.ir.inference.inference_pass_test import (
     ImportError: No module named ir.inference.inference_pass_test
```
No other CI configuration complained about this import statement, not saying that locally it is working for me. 

The PR-CI-Py3 and PR-CI-Windows (which i re-run) failures where caused because they tried to run cpu tests on GPU configuration (test_batch_fc_op)

File: paddle/fluid/platform/device_context.cc [hunk 2]
```diff
@@ -502,6 +504,7 @@ MKLDNNDeviceContext::MKLDNNDeviceContext(CPUPlace place)
     : CPUDeviceContext(place), p_blobmap_() {
   p_blobmap_.reset(new BlobMap());
   p_mutex_.reset(new std::mutex());
+  if (FLAGS_onednn_use_input_mem_format) use_input_mem_format_ = true;
```
Comment (jczaja) [inline paddle/fluid/platform/device_context.cc:507]: why not use: __use_input_mem_format = FLAGS_onednn_use_input_mem_format;__  ?

File: paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc [hunk 3]
```diff
@@ -206,13 +206,23 @@ class ConvMKLDNNHandlerT
...
-          platform::MKLDNNMemDesc(src_tz, data_type, chosen_memory_format);
-      const auto weights_md = platform::MKLDNNMemDesc(weights_tz, data_type,
-                                                      MKLDNNMemoryFormat::any);
-      const auto dst_md = platform::MKLDNNMemDesc(
+      auto dst_md = platform::MKLDNNMemDesc(
           dst_tz, platform::MKLDNNGetDataType<T_out>(), chosen_memory_format);
 
+      platform::MKLDNNMemoryDescriptor src_md, weights_md;
+
+      if (this->dev_ctx_.UseInputMemFormat()) {
```
Comment (jczaja) [inline paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc:214]: Same logic  should be made for ComputeInt8 and ConvMKLDNNGrad::Compute()

File: paddle/fluid/platform/device_context.cc (hunk 2 above)
Comment (arogowie-intel) [reply paddle/fluid/platform/device_context.cc:507]: Good  idea.

**Pull request closed without merging**